internals/search_fulltext.py

```diff
@@ -0,0 +1,216 @@
+
+  strings.append(fe.motivation)
+  strings.append(fe.devtrial_instructions)
+  strings.append(fe.activation_risks)
+  strings.append(fe.measurement)
+
+  strings.append(fe.initial_public_proposal_url)
+  strings.extend(fe.explainer_links)
+  # TODO: standard_maturity
+  strings.append(fe.spec_link)
+  strings.extend(fe.spec_mentor_emails)
+  strings.append(fe.interop_compat_risks)
+  strings.append(fe.all_platforms_descr)
+  strings.append(fe.tag_review)
+  # TODO: tag_review_status
+  strings.append(fe.non_oss_deps)
+  strings.append(fe.anticipated_spec_changes)
+
+  # TODO: ff_views
+  # TODO: safari_views
+  # TODO: web_dev_views
+  strings.append(fe.ff_views_link)
+  strings.append(fe.safari_views_link)
+  strings.append(fe.web_dev_views_link)
+  strings.append(fe.ff_views_notes)
+  strings.append(fe.safari_views_notes)
+  strings.append(fe.web_dev_views_notes)
+  strings.append(fe.other_views_notes)
+
+  strings.append(fe.security_risks)
+  # TODO: security_review_status
+  # TODO: privacy_review_status
+
+  strings.append(fe.ergonomics_risks)
+  strings.append(fe.wpt_descr)
+  strings.append(fe.webview_risks)
+
+  strings.extend(fe.devrel_emails)
+  strings.append(fe.debuggability)
+  strings.extend(fe.doc_links)
+  strings.extend(fe.sample_links)
+
+  return strings
+
+
+def parse_words(strings):
+  """Return a list of all searchable words in the given feature entry."""
+  words = set()
+  for s in strings:
+    # skip missing fields and eliminate apostrophes
+    if not s: continue
+    s = s.lower().replace("'", "")
+    words.update(WORD_RE.findall(s))
+
+  words.difference_update(STOP_WORDS)
+  return words
+
+
+def batch_index_features(fe_list, existing_fw_list):
+  """XXX."""
+  existing_fw_dict = {
+      fw.feature_id: fw
+      for fw in existing_fw_list}
+  updated_fw_list = []
+
+  for fe in fe_list:
+    feature_id = fe.key.integer_id()
+    feature_words = (existing_fw_dict.get(feature_id) or
+                     FeatureWords(feature_id=feature_id))
+
+    words = parse_words(get_strings(fe))
+    logging.info('feature %r has words %r', feature_id, words)
+    feature_words.words = words
+    updated_fw_list.append(feature_words)
+
+  return updated_fw_list
+
+
+def index_feature(fe):
+  """Create or update a word bag for the given feature entry."""
+  feature_id = fe.key.integer_id()
+  query = FeatureWords.query(FeatureWords.feature_id == feature_id)
+  existing_fw_list = query.fetch(None)
+  updated_fw_list = batch_index_features([fe], existing_fw_list)
+  updated_fw_list[0].put()
+
+
+def search_fulltext(textterm):
+  """Return IDs of features that have some word(s) from phrase."""
+  search_words = parse_words([textterm])
+  if not search_words:
+    logging.warn('Cannot process fulltext term: %r', textterm)
+    return []  # user is searching for stop words.
+
+  logging.info('looking for words: %r', search_words)
+  query = FeatureWords.query(FeatureWords.words.IN(search_words))
```

Comment:
What is the result of this? Does it return any FeatureWords entity that contains every word specified in the search?

Reply:
It will match any FeatureWords entity that has any word in search_words.
Which is wrong... we want only those that have every word.  I will change that.